File: typescript/tests/elderly_board.test.ts
// elderly_board.test.ts
// Verifies the CircleAI.Elderly port: per-resident care plans, medication
// reminders (activate/deactivate), latest check-in, and missed-check-in test.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryElderlyCareBoard,
  ElderlyDomainContext,
  carePlan,
  medReminder,
  checkIn,
} from "../src/elderly/index";

const D = (s: string) => new Date(s);

describe("InMemoryElderlyCareBoard", () => {
  it("stores a care plan keyed by resident name", () => {
    const b = new InMemoryElderlyCareBoard();
    b.setPlan(carePlan("cp1", "Gogo", ["Hypertension"], ["Penicillin"], "Prefers morning walks"));
    assert.equal(b.getPlan("Gogo")?.carerNotes, "Prefers morning walks");
    assert.deepEqual(b.getPlan("Gogo")?.medicalConditions, ["Hypertension"]);
    assert.equal(b.getPlan("Nobody"), undefined);
  });

  it("filters active reminders per resident and deactivates by id", () => {
    const b = new InMemoryElderlyCareBoard();
    b.addReminder(medReminder("r1", "Gogo", "Amlodipine", 8 * 3600_000, true));
    b.addReminder(medReminder("r2", "Gogo", "Aspirin", 20 * 3600_000, true));
    b.addReminder(medReminder("r3", "Mkhulu", "Metformin", 8 * 3600_000, true));
    assert.deepEqual(
      b.activeRemindersFor("Gogo").map((r) => r.reminderId),
      ["r1", "r2"],
    );
    b.deactivateReminder("r1");
    assert.deepEqual(
      b.activeRemindersFor("Gogo").map((r) => r.reminderId),
      ["r2"],
    );
    assert.throws(() => b.deactivateReminder("ghost"), /Unknown reminder ghost/);
  });

  it("returns the latest check-in and detects missed check-ins", () => {
    const b = new InMemoryElderlyCareBoard();
    assert.equal(b.latestCheckIn("Gogo"), undefined);
    assert.equal(b.missedCheckIn("Gogo", D("2026-05-01T00:00:00Z")), true); // no check-in at all
    b.recordCheckIn(checkIn("k1", "Gogo", D("2026-05-01T08:00:00Z"), "OK", null));
    b.recordCheckIn(checkIn("k2", "Gogo", D("2026-05-02T08:00:00Z"), "OK", "cheerful"));
    assert.equal(b.latestCheckIn("Gogo")?.checkInId, "k2");
    assert.equal(b.missedCheckIn("Gogo", D("2026-05-02T00:00:00Z")), false); // latest after cutoff
    assert.equal(b.missedCheckIn("Gogo", D("2026-05-03T00:00:00Z")), true); // latest before cutoff
  });

  it("rejects null arguments", () => {
    const b = new InMemoryElderlyCareBoard();
    assert.throws(() => b.setPlan(null as never));
    assert.throws(() => b.addReminder(null as never));
    assert.throws(() => b.recordCheckIn(null as never));
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(ElderlyDomainContext.systemPromptSnippet.includes("[DOMAIN: Elderly]"));
    assert.deepEqual(ElderlyDomainContext.complianceFlags, ["Older_Persons_Act_13_2006", "Social_Assistance_Act", "POPIA"]);
    assert.deepEqual(ElderlyDomainContext.suggestedTools, ["medication_reminder", "calendar", "web_search", "document_editor"]);
  });
});

File: go/tests/elderly_board_test.go
// elderly_board_test.go
//
// Verifies the CircleAI.Elderly port (elderly_board.go): care-plan set/get
// (defensive list copy), reminder add/deactivate (unknown-id error) +
// ActiveRemindersFor (resident + active filter), check-in record + LatestCheckIn
// (most recent; absent when none), and MissedCheckIn (no check-in or latest before
// `since`).

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestElderly_PlanAndReminders(t *testing.T) {
	b := circleai.NewInMemoryElderlyCareBoard()
	conditions := []string{"Diabetes"}
	b.SetPlan(circleai.CarePlan{PlanId: "cp1", ResidentName: "Gogo", MedicalConditions: conditions, Allergies: []string{"Penicillin"}, CarerNotes: "Gentle"})
	conditions[0] = "MUTATED" // must not affect stored plan

	plan, ok := b.GetPlan("Gogo")
	if !ok || len(plan.MedicalConditions) != 1 || plan.MedicalConditions[0] != "Diabetes" {
		t.Fatalf("get plan wrong (defensive copy?): %+v ok=%v", plan, ok)
	}
	if _, ok := b.GetPlan("Nobody"); ok {
		t.Fatalf("unknown resident plan must be absent")
	}

	b.AddReminder(circleai.MedReminder{ReminderId: "r1", ResidentName: "Gogo", Medication: "Metformin", DailyAt: 8 * time.Hour, Active: true})
	b.AddReminder(circleai.MedReminder{ReminderId: "r2", ResidentName: "Gogo", Medication: "Aspirin", DailyAt: 20 * time.Hour, Active: true})
	b.AddReminder(circleai.MedReminder{ReminderId: "r3", ResidentName: "Mkhulu", Medication: "Statin", DailyAt: 21 * time.Hour, Active: true})

	if err := b.DeactivateReminder("r2"); err != nil {
		t.Fatalf("deactivate: %v", err)
	}
	if err := b.DeactivateReminder("ghost"); err == nil {
		t.Fatalf("unknown reminder deactivate must error")
	}
	active := b.ActiveRemindersFor("Gogo")
	// r1 active (r2 deactivated, r3 is a different resident). Sorted by ReminderId.
	if len(active) != 1 || active[0].ReminderId != "r1" {
		t.Fatalf("active reminders wrong: %+v", active)
	}
}

func TestElderly_CheckIns(t *testing.T) {
	b := circleai.NewInMemoryElderlyCareBoard()
	t0 := time.Date(2026, 7, 1, 8, 0, 0, 0, time.UTC)
	b.RecordCheckIn(circleai.ElderlyCheckIn{CheckInId: "k1", ResidentName: "Gogo", AtUtc: t0, Status: "OK"})
	b.RecordCheckIn(circleai.ElderlyCheckIn{CheckInId: "k2", ResidentName: "Gogo", AtUtc: t0.Add(12 * time.Hour), Status: "OK"})

	latest, ok := b.LatestCheckIn("Gogo")
	if !ok || latest.CheckInId != "k2" {
		t.Fatalf("latest check-in = %+v ok=%v, want k2", latest, ok)
	}
	if _, ok := b.LatestCheckIn("Nobody"); ok {
		t.Fatalf("unknown resident latest check-in must be absent")
	}

	// Not missed: latest (t0+12h) is at/after `since` = t0.
	if b.MissedCheckIn("Gogo", t0) {
		t.Fatalf("should not be missed when latest >= since")
	}
	// Missed: `since` is after the latest check-in.
	if !b.MissedCheckIn("Gogo", t0.Add(24*time.Hour)) {
		t.Fatalf("should be missed when latest < since")
	}
	// Missed: no check-in at all.
	if !b.MissedCheckIn("Nobody", t0) {
		t.Fatalf("should be missed when no check-in exists")
	}
}
